### PR TITLE
Edit TE UI

### LINKED ISSUE #1144: Edit start/stop time UI
- [x] A dialog picker is used to select a single time (hours:minutes).
- [x] The selected time is indicated by the filled circle at the end of the clock hand.
- [x] A time picker adjusts to a user’s preferred time setting, i.e. the 12-hour or 24-hour format.
- [x] A user can tap between the hour and minute label in the title bar to switch between the hour or minute view. Additionally, a user can tap on the AM/PM label in the title bar to switch between AM or PM (in 12-hour clock mode).

### Start/stop time edit view
edit - time

Zeplin: https://zpl.io/blrBrn1

### LINKED ISSUE #1145: Edit date UI
- [x] A dialog picker is used to select a single date.
- [x] The selected day is indicated by a filled circle. The current day is indicated by a different colour and type weight.
- [x] A user can swipe left to right to navigate through the months.

### Date edit view
edit - date

Zeplin: https://zpl.io/2ZqdqAJ

### PR DESCRIPTION
**Pull request by kornelijepetak**
Ref #1121 
Closes #1144 
Closes #1145 

There are several things that need to be fixed, but I was unable to find a good solution.
Some of the things also need design (behaviour, that is) better defined.

This PR goes only to introduce the edit view so that at least one can see the time entry.
Related views (tag, project etc. selectors) will come as part of other PRs.

### CODE REVIEW

**Comment by willsb** [comment]
This is on hold until the new `SelectDateTimeViewModel` is merged

**Comment by kornelijepetak** [comment]
This is ready for review.

In the end, I complicated more with separating PRs than necessary and I decided not to waste precious time on that so I hope you will be able to review this as is. I think the PR isn't that complicated, basically there are only three (rather isolated) things requiring any exercise:

* DurationEditText implementation (for editing durations)
* TextViewTagListTargetBinding (for creating tag tokens)
* DateTimeDialogFragment (called for editing dates and times)

Everything else is a breeze.

**Notes for for custom DurationEditText**
The issue with duration editing is that soft input events can't be relied upon, so I have utilized the `InputFilter` to help with this. Otherwise, very complex text editing logic would need to be implemented, or a surrogate view would have to be used.

**Notes for for custom TextViewTagListTargetBinding**
The thing about this is that it still doesn't do ellipsis if there are too many tags or if they are too long. I wasn't able to solve this yet.

**Comment by kornelijepetak** [comment]
I forgot to add that this also closes #1145

File: Toggl.Giskard/Activities/EditTimeEntryActivity.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+            toolbar.NavigationClick += onNavigateBack;
+        }
+
+        private void onNavigateBack(object sender, NavigationClickEventArgs e)
+        {
+            ViewModel.CloseCommand.Execute();
+        }
+
+        public override void Finish()
```
Comment (willsb) [inline Toggl.Giskard/Activities/EditTimeEntryActivity.cs]: No need to override this

File: Toggl.Giskard/Activities/EditTimeEntryActivity.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            base.Finish();
+        }
+
+        protected override void Dispose(bool disposing)
```
Comment (willsb) [inline Toggl.Giskard/Activities/EditTimeEntryActivity.cs]: Or this

File: Toggl.Giskard/Fragments/DateTimeDialogFragment.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+using Android.Text.Format;
+using MvvmCross.Droid.Support.V4;
+using MvvmCross.Droid.Views.Attributes;
+using Toggl.Foundation.MvvmCross.Parameters;
+using Toggl.Foundation.MvvmCross.ViewModels;
+
+namespace Toggl.Giskard.Fragments
+{
+    [MvxDialogFragmentPresentation(AddToBackStack = true)]
+    public class DateTimeDialogFragment : MvxDialogFragment<SelectDateTimeViewModel>
```
Comment (willsb) [inline Toggl.Giskard/Fragments/DateTimeDialogFragment.cs]: This should be named `SelectDateTimeFragment`, please

File: Toggl.Giskard/Fragments/DateTimeDialogFragment.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+            dialog.DatePicker.MinDate = ViewModel.MinDate.ToUnixTimeMilliseconds();
+            dialog.DatePicker.MaxDate = ViewModel.MaxDate.ToUnixTimeMilliseconds();
+
+            return dialog;
+        }
+
+        private TimePickerDialog showTime()
+        {
+            var is24hourFormat = DateFormat.Is24HourFormat(Activity);
```
Comment (willsb) [inline Toggl.Giskard/Fragments/DateTimeDialogFragment.cs]: We should check our own setting instead. This should be included in the VM, if it already isn't

File: Toggl.Giskard/Fragments/DateTimeDialogFragment.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        }
+
+        private void onTimeSet(object sender, TimePickerDialog.TimeSetEventArgs e)
+        {
+            dateTime = new DateTimeOffset(ViewModel.CurrentDateTime.Year, ViewModel.CurrentDateTime.Month, ViewModel.CurrentDateTime.Day,
+                                          e.HourOfDay, e.Minute, ViewModel.CurrentDateTime.Second,
+                                          TimeSpan.Zero);
+        }
+
+        public override void OnCancel(Android.Content.IDialogInterface dialog)
```
Comment (willsb) [inline Toggl.Giskard/Fragments/DateTimeDialogFragment.cs]: Add `Android.Content ` to the using list

File: Toggl.Giskard/LinkerPleaseInclude.cs [hunk 6]
```diff
@@ -16,6 +16,8 @@ namespace Toggl.Giskard
     // are preserved in the deployed app
     public class LinkerPleaseInclude
     {
+        private MvxCompoundButtonCheckedTargetBinding binding = new MvxCompoundButtonCheckedTargetBinding(null, null);
```
Comment (willsb) [inline Toggl.Giskard/LinkerPleaseInclude.cs]: Please use the Include methods for consistency

File: Toggl.Giskard/Resources/layout/EditLayout.axml [hunk 7]
```diff
@@ -0,0 +1,316 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:tools="http://schemas.android.com/tools"
+    xmlns:local="http://schemas.android.com/apk/res-auto"
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:focusable="true"
+    android:orientation="vertical"
+    android:focusableInTouchMode="true"
```
Comment (willsb) [inline Toggl.Giskard/Resources/layout/EditLayout.axml]: Why is this needed?

File: Toggl.Giskard/Resources/layout/EditLayout.axml [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+        android:id="@+id/EditBillableLabel"
+        android:textSize="15sp"
+        android:text="@string/Billable"
+        android:gravity="center_vertical"
+        android:textColor="@color/darkLabel"
+        android:layout_height="48dp"
+        android:layout_width="match_parent"
+        android:layout_alignTop="@id/EditBillableIcon"
+        android:layout_toRightOf="@id/EditBillableIcon" />
+    <android.support.v7.widget.SwitchCompat
```
Comment (willsb) [inline Toggl.Giskard/Resources/layout/EditLayout.axml]: Is this needed? What for?

File: Toggl.Giskard/Resources/layout/EditLayout.axml [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+        android:layout_alignTop="@id/EditBillableIcon" />
+    <View
+        android:background="@color/separator"
+        android:layout_height="0.5dp"
+        android:layout_marginLeft="56dp"
+        android:layout_width="match_parent"
+        android:layout_below="@id/EditBillableLabel" />
+    <LinearLayout
+        android:id="@+id/EditTimeValuesContainer"
+        local:MvxBind="Click SelectStartTimeCommand"
```
Comment (willsb) [inline Toggl.Giskard/Resources/layout/EditLayout.axml]: Shouldn't this be in the relative layout below instead?

File: Toggl.Giskard/Resources/layout/EditLayout.axml [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+        android:layout_marginRight="18dp"
+        android:layout_alignParentRight="true"
+        android:layout_alignTop="@id/EditBillableIcon" />
+    <View
+        android:background="@color/separator"
+        android:layout_height="0.5dp"
+        android:layout_marginLeft="56dp"
+        android:layout_width="match_parent"
+        android:layout_below="@id/EditBillableLabel" />
+    <LinearLayout
```
Comment (willsb) [inline Toggl.Giskard/Resources/layout/EditLayout.axml]: This Linear Layout is not necessary

File: Toggl.Giskard/Resources/layout/EditLayout.axml [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+        android:layout_below="@+id/EditBillableIcon" />
+    <View
+        android:id="@+id/EditTimesBottomBorder"
+        android:background="@color/separator"
+        android:layout_height="0.5dp"
+        android:layout_marginTop="72dp"
+        android:layout_marginLeft="56dp"
+        android:layout_width="match_parent"
+        android:layout_below="@id/EditBillableLabel" />
+    <Toggl.Giskard.Views.DurationEditText
```
Comment (willsb) [inline Toggl.Giskard/Resources/layout/EditLayout.axml]: `toggl.giskard.views.DurationEditText`

File: Toggl.Giskard/Resources/layout/EditLayout.axml [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+        android:layout_width="match_parent"
+        android:layout_below="@id/EditDateValueLabel" />
+    <android.support.v7.widget.Toolbar
+        android:id="@+id/Toolbar"
+        android:elevation="4dp"
+        android:minHeight="?attr/actionBarSize"
+        android:theme="@style/ThemeOverlay.AppCompat.ActionBar"
+        android:layout_height="56dp"
+        android:layout_width="match_parent">
+        <TextView
```
Comment (willsb) [inline Toggl.Giskard/Resources/layout/EditLayout.axml]: I'm 100% against using text in the toolbar. Heidi will look into changing this next week

File: Toggl.Giskard/Fragments/DateTimeDialogFragment.cs (hunk 4 above)
Comment (kornelijepetak) [reply Toggl.Giskard/Fragments/DateTimeDialogFragment.cs]: As far as I know, we don't have this in our data source yet. There is also another issue open for this (#1349). And this PR is already big as it is.

I have for now set this to take a hardcoded value from the viewmodel. 
If you have a better solution at the moment, please propose.

File: Toggl.Giskard/Resources/layout/EditLayout.axml (hunk 7 above)
Comment (kornelijepetak) [reply Toggl.Giskard/Resources/layout/EditLayout.axml]: Because otherwise, when you click 'Done' on the soft input when editing duration, the next available focusable view gets focus and that is `description`. This way (hack) the root layout itself gets focus effectively clearing the focus from the duration. I tried multiple ways, and this is the only one that worked. Please advise if it can be done some other way.

File: Toggl.Giskard/Resources/layout/EditLayout.axml (hunk 9 above)
Comment (kornelijepetak) [reply Toggl.Giskard/Resources/layout/EditLayout.axml]: Not sure which part you mean. What should be where?

File: Toggl.Giskard/Resources/layout/EditLayout.axml (hunk 10 above)
Comment (kornelijepetak) [reply Toggl.Giskard/Resources/layout/EditLayout.axml]: IIRC I couldn't make weight work correctly for the inner views to take 50/50 space... Will try again.

File: Toggl.Giskard/Resources/layout/EditLayout.axml (hunk 9 above)
Comment (kornelijepetak) [reply Toggl.Giskard/Resources/layout/EditLayout.axml]: Oh, I get it. True. It should be in the relative layout below.

File: Toggl.Giskard/Resources/layout/EditLayout.axml (hunk 10 above)
Comment (kornelijepetak) [reply Toggl.Giskard/Resources/layout/EditLayout.axml]: I've tried to fix it now without the container layout and can't find a way to do that. `layout_weight` doesn't work in `RelativeLayout`.

File: Toggl.Daneel/Views/EditDuration/DurationField.cs [hunk 13]
```diff
@@ -108,17 +108,17 @@ private void finishEditing(object sender, EventArgs e)
 
         private void tryUpdate(DurationFieldInfo nextInput)
         {
-            if (nextInput.Equals(input) == false)
-            {
-                input = nextInput;
+            if (nextInput.Equals(input))
```
Comment (simonrozsival) [inline Toggl.Daneel/Views/EditDuration/DurationField.cs]: Why do you update Daneel in this PR?

File: Toggl.Giskard/Activities/EditTimeEntryActivity.cs [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+            toolbar.NavigationClick += onNavigateBack;
+        }
+
+        private void onNavigateBack(object sender, NavigationClickEventArgs e)
+        {
+            ViewModel.CloseCommand.Execute();
+        }
+    }
+
+
```
Comment (simonrozsival) [inline Toggl.Giskard/Activities/EditTimeEntryActivity.cs]: Extra new lines

File: Toggl.Giskard/Views/DurationEditText.cs [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+                    int digit = int.Parse(lastChar.ToString());
+                    onDigit?.Invoke(digit);
+
+                    return new Java.Lang.String(digit.ToString());
+                }
+
+                return empty;
+            }
+        }
+
```
Comment (simonrozsival) [inline Toggl.Giskard/Views/DurationEditText.cs]: Extra newline

File: Toggl.Giskard/Views/DurationEditText.cs [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+            SetFilters(new IInputFilter[] { filter });
+        }
+
+        private void showCurrentDuration()
+        {
+            var text = (string)converter?.Convert(duration, typeof(TimeSpan), null, CultureInfo.CurrentCulture) ?? string.Empty;
+            setText(text);
+        }
+
+        private void setText(string text)
```
Comment (simonrozsival) [inline Toggl.Giskard/Views/DurationEditText.cs]: I think this private method is not necessary at all - in the iOS version I have to convert `string` into `NSAttributedString` and so I added that method, but I don't see why it would be beneficial in this case.

File: Toggl.Giskard/Resources/layout/EditLayout.axml (hunk 12 above)
Comment (kornelijepetak) [reply Toggl.Giskard/Resources/layout/EditLayout.axml]: This was confirmed by @pe0ny. For now we go without.
